### PR TITLE
homescreen nav api (bug 1214745)

### PR DESCRIPTION
**Pull request by washort**

### CODE REVIEW

File: mkt/search/tests/test_views.py [hunk 1]
```diff
@@ -1447,22 +1460,30 @@ def test_statsd(self, _mock):
         assert _mock.called
 
     def test_search_no_doc_type_passed(self):
+        self.makeHomescreen()
```
Comment (diox) [inline mkt/search/tests/test_views.py]: probably best to make a separate test for this IMHO

File: mkt/search/tests/test_views.py [hunk 2]
```diff
@@ -1507,12 +1528,13 @@ def test_search_popularity(self):
         eq_(objs[2]['slug'], self.webapp.app_slug)
 
     def test_search_sort_by_reviewed(self):
+        self.makeHomescreen()
```
Comment (diox) [inline mkt/search/tests/test_views.py]: same as above, a separate test is probably better, those tests depend on relevancy/reviewed time, so they are quite fragile. your call though.

File: mkt/search/filters.py [hunk 3]
```diff
@@ -162,8 +162,9 @@ def filter_queryset(self, request, queryset, view):
 
 class PublicSearchFormFilter(SearchFormFilter):
     VALID_FILTERS = ['app_type', 'author.raw', 'category', 'device', 'guid',
-                     'installs_allowed_from', 'is_offline', 'manifest_url',
```
Comment (diox) [inline mkt/search/filters.py]: I don't see a change here

File: mkt/search/views.py [hunk 4]
```diff
@@ -94,6 +94,7 @@ class MultiSearchView(SearchView):
     serializer_classes = {
         'extension': ESExtensionSerializer,
         'webapp': ESAppSerializer,
+        'homescreen': ESAppSerializer,
```
Comment (diox) [inline mkt/search/views.py]: wait what ? this is not a different type, so this should not be needed

File: mkt/search/tests/test_views.py [hunk 5]
```diff
@@ -1549,6 +1577,20 @@ def test_search_q(self):
         eq_(objs[2]['name'], self.webapp.name)
         eq_(objs[2]['slug'], self.webapp.app_slug)
 
+    def test_homescreen(self):
+        homescreen = self.makeHomescreen()
+        res = self.anon.get(self.url, data={
+            'doc_type': 'extension,webapp,website', 'lang': 'en-US',
+            'is_homescreen': '1'})
```
Comment (diox) [inline mkt/search/tests/test_views.py]: I think you're missing a test for when it's set to 0

Comment (diox) [inline mkt/search/tests/test_views.py]: you could have such a test in `test_filters` actually that'd be better

File: mkt/webapps/indexers.py [hunk 6]
```diff
@@ -297,6 +297,7 @@ def extract_document(cls, pk=None, obj=None):
         d['is_rereviewed'] = is_rereviewed
         d['rereview_date'] = (obj.rereviewqueue_set.get().created
                               if is_rereviewed else None)
+        d['is_homescreen'] = obj.is_homescreen()
```
Comment (ngokevin) [inline mkt/webapps/indexers.py]: add this to the mapping

File: mkt/search/tests/test_views.py (hunk 1 above)
Comment (ngokevin) [inline mkt/search/tests/test_views.py]: +1

File: mkt/search/tests/test_views.py [hunk 7]
```diff
@@ -1404,6 +1404,19 @@ def setUp(self):
             reviewed=self.days_ago(0), status=mkt.STATUS_PUBLIC)
         self.refresh(('webapp', 'website', 'extension'))
 
+    def makeHomescreen(self):
```
Comment (ngokevin) [inline mkt/search/tests/test_views.py]: underscore

File: mkt/search/filters.py [hunk 8]
```diff
@@ -204,6 +205,23 @@ class ExtensionSearchFormFilter(SearchFormFilter):
     VALID_FILTERS = ['author.raw']
 
 
+class HomescreenFilter(BaseFilterBackend):
+    """
+    A django-rest-framework filter backend that filters in or out homescreens.
+    """
+    def filter_queryset(self, request, queryset, view):
+        is_homescreen = request.GET.get('is_homescreen')
```
Comment (ngokevin) [inline mkt/search/filters.py]: maybe we should check webapp_type === 'homescreen' rather than is_homescreen === '0'

**Comment by ngokevin** [comment]
update for re-r

File: mkt/search/filters.py (hunk 8 above)
Comment (washort) [inline mkt/search/filters.py]: explain the difference to me?

File: mkt/webapps/serializers.py [hunk 9]
```diff
@@ -530,6 +534,9 @@ def get_package_path(self, obj):
     def get_file_size(self, obj):
         return obj.es_data.get('file_size')
 
+    def get_is_homescreen(self, obj):
+        return obj.es_data['is_homescreen']
```
Comment (diox) [inline mkt/webapps/serializers.py]: this will fail before a reindex is made. use `.get('is_homescreen')`

File: mkt/search/filters.py (hunk 8 above)
Comment (ngokevin) [inline mkt/search/filters.py]: seems to me it's better api design. have one filter webapp_type=whatever rather than is_this and is_that

Comment (washort) [inline mkt/search/filters.py]: ah, I see what you mean. fireplace is already using `is_homescreen` so let's fix it after next week

Comment (ngokevin) [inline mkt/search/filters.py]: it's not using it, we can just change it here

File: mkt/search/tests/test_views.py [hunk 10]
```diff
@@ -1525,6 +1564,34 @@ def test_search_sort_by_reviewed(self):
...
+
+    def test_search_homescreen_sort_by_reviewed(self):
+        self.makeHomescreen()
+        res = self.anon.get(self.url, data={
+            'doc_type': 'extension,webapp,website', 'lang': 'en-US',
+            'sort': 'reviewed'})
+        eq_(res.status_code, 200)
+        objs = res.json['objects']
+        eq_(len(objs), 4)
+        eq_(objs[0]['doc_type'], 'extension')
```
Comment (ngokevin) [inline mkt/search/tests/test_views.py]: if testing sort, i imagine just testing the ID is sufficient

**Comment by ngokevin** [comment]
r+wc

**Pull request merged**